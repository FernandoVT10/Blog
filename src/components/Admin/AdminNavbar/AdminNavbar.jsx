import ApiController from "@/services/ApiController";

import Link from "next/link";
import { useEffect, useState } from "react";

import "./AdminNavbar.scss";

function AdminNavbar() {
    const [navbarStatus, setNavbarStatus] = useState(false);
    const [pendingMessagesCount, setPendingMessagesCount] = useState(0);

    useEffect(() => {
        ApiController.get("messages/pendingMessages", true)
        .then(res => {
            if(res.data) {
                setPendingMessagesCount(res.data.count);
            }
        });
    }, []);

    const navbarClass = navbarStatus ? "admin-navbar--active" : "";

    return (
        <nav className={`admin-navbar ${navbarClass}`}>
            <ul className="admin-navbar__items-container">
                <li className="admin-navbar__item-header">
                    <span
                    className="admin-navbar__item-icon"
                    onClick={() => setNavbarStatus(!navbarStatus)}>
                        <i className="fas fa-bars" aria-hidden="true"></i>
                    </span>

                    <h4 className="admin-navbar__item-title">Admin Blog</h4>
                </li>
                <li className="admin-navbar__item">
                    <Link href="/">
                        <a>
                            <span className="admin-navbar__item-icon">
                                <i className="fas fa-globe-americas" aria-hidden="true"></i>
                            </span>

                            <span className="admin-navbar__item-text">
                                Back to Website
                            </span>
                        </a>
                    </Link>
                </li>
                <li className="admin-navbar__item">
                    <Link href="/admin">
                        <a>
                            <span className="admin-navbar__item-icon">
                                <i className="fas fa-th-large" aria-hidden="true"></i>
                            </span>

                            <span className="admin-navbar__item-text">
                                Dashboard
                            </span>
                        </a>
                    </Link>
                </li>
                <li className="admin-navbar__item">
                    <Link href="/admin/articles">
                        <a>
                            <span className="admin-navbar__item-icon">
                                <i className="fas fa-images" aria-hidden="true"></i>
                            </span>

                            <span className="admin-navbar__item-text">
                                Articles
                            </span>
                        </a>
                    </Link>
                </li>
                <li className="admin-navbar__item">
                    <Link href="/admin/projects">
                        <a>
                            <span className="admin-navbar__item-icon">
                                <i className="fas fa-folder-open" aria-hidden="true"></i>
                            </span>

                            <span className="admin-navbar__item-text">
                                Projects
                            </span>
                        </a>
                    </Link>
                </li>
                <li className="admin-navbar__item">
                    <Link href="/admin/messages">
                        <a>
                            <span className="admin-navbar__item-icon">
                                <i className="fas fa-envelope" aria-hidden="true"></i>

                                { pendingMessagesCount &&
                                    <span className="admin-navbar__item-badge">
                                        { pendingMessagesCount }
                                    </span>
                                }
                            </span>

                            <span className="admin-navbar__item-text">
                                Messages
                            </span>
                        </a>
                    </Link>
                </li>
            </ul>
        </nav>
    );
}

export default AdminNavbar;