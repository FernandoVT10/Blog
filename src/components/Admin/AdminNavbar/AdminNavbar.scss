.admin-navbar {
    position: fixed;
    z-index: 1000;
    top: 0;
    left: 0;
    background: var(--primary);
    width: 50px;
    height: 100vh;
    overflow-x: hidden;
    overflow-y: auto;
    transition: .3s;

    &--active {
        width: 220px;
    }

    &__items-container {
        display: block;
        list-style: none;
        padding: 0;
        margin: 0;
    }

    &__item-header {
        position: relative;
        padding: 0;
        margin: 0;
    }

    &__item {
        cursor: pointer;
        position: relative;
        padding: 0;
        margin: 0;
        transition: .3s;

        &:hover {
            background: var(--secondary);
        }
    }

    &__item-icon {
        cursor: pointer;
        display: flex;
        justify-content: center;
        align-items: center;
        width: 50px;
        height: 50px;
        font-size: 20px;
        color: var(--white);
        transition: .3s;

        &:hover {
            background: var(--secondary);
        }
    }

    &__item-title {
        position: absolute;
        white-space: nowrap;
        top: 13px;
        left: 65px;
        color: var(--white);
        font-size: 18px;
        font-weight: bold;
    }

    &__item-text {
        position: absolute;
        top: 15px;
        white-space: nowrap;
        left: 65px;
        font-size: 14px;
        color: darken($color: #ffffff, $amount: 20);
        font-weight: bold;
        transition: .3s;
    }

    &__item:hover &__item-text {
        color: var(--white);
    }

    &__item-badge {
        position: absolute;
        top: 8px;
        left: 25px;
        display: flex;
        justify-content: center;
        color: var(--black);
        font-size: 14px;
        font-weight: bold;
        width: 20px;
        height: 20px;
        border-radius: 50%;
        background: var(--secondary);
    }

    @media screen and (max-width: 768px) {
        top: auto;
        width: 100%;
        height: 50px;
        bottom: 0;

        &__items-container {
            display: flex;
        }

        &__item-header {
            display: none;
        }

        &__item {
            width: 100%;
        }

        &__item-icon {
            width: 100%;
        }

        &__item-text,
        &__item-title {
            display: none;
        }

        &__item-badge {
            left: 50%;
            right: 0;
        }
    }
}