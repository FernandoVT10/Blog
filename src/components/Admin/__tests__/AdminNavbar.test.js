import { act, Simulate } from "react-dom/test-utils";
import { render } from "react-dom";
import AdminNavbar from "../AdminNavbar/";

let container;

describe('<AdminNavbar/> Component', () => {
    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });
    
    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it('Check if renders correctly', () => {
        render(<AdminNavbar/>, container);

        expect(container).toMatchSnapshot();
    });

    it("It should add the navbar--active class", async () => {
        render(<AdminNavbar/>, container);

        const nav = container.querySelector("nav");
        const button = container.querySelector(".admin-navbar__item-icon");

        expect(nav.classList.contains("admin-navbar--active")).toBeFalsy();

        act(() => {
            Simulate.click(button);
        });

        expect(nav.classList.contains("admin-navbar--active")).toBeTruthy();
    });
});