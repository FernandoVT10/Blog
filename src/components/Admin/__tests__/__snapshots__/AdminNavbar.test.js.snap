// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminNavbar/> Component Check if renders correctly 1`] = `
<div>
  <nav
    class="admin-navbar "
  >
    <ul
      class="admin-navbar__items-container"
    >
      <li
        class="admin-navbar__item-header"
      >
        <span
          class="admin-navbar__item-icon"
        >
          <i
            aria-hidden="true"
            class="fas fa-bars"
          />
        </span>
        <h4
          class="admin-navbar__item-title"
        >
          Admin Blog
        </h4>
      </li>
      <li
        class="admin-navbar__item"
      >
        <a
          href="/"
        >
          <span
            class="admin-navbar__item-icon"
          >
            <i
              aria-hidden="true"
              class="fas fa-globe-americas"
            />
          </span>
          <span
            class="admin-navbar__item-text"
          >
            Back to Website
          </span>
        </a>
      </li>
      <li
        class="admin-navbar__item"
      >
        <a
          href="/admin"
        >
          <span
            class="admin-navbar__item-icon"
          >
            <i
              aria-hidden="true"
              class="fas fa-th-large"
            />
          </span>
          <span
            class="admin-navbar__item-text"
          >
            Dashboard
          </span>
        </a>
      </li>
      <li
        class="admin-navbar__item"
      >
        <a
          href="/admin/articles"
        >
          <span
            class="admin-navbar__item-icon"
          >
            <i
              aria-hidden="true"
              class="fas fa-images"
            />
          </span>
          <span
            class="admin-navbar__item-text"
          >
            Articles
          </span>
        </a>
      </li>
      <li
        class="admin-navbar__item"
      >
        <a
          href="/admin/projects"
        >
          <span
            class="admin-navbar__item-icon"
          >
            <i
              aria-hidden="true"
              class="fas fa-folder-open"
            />
          </span>
          <span
            class="admin-navbar__item-text"
          >
            Projects
          </span>
        </a>
      </li>
      <li
        class="admin-navbar__item"
      >
        <a
          href="/admin/messages"
        >
          <span
            class="admin-navbar__item-icon"
          >
            <i
              aria-hidden="true"
              class="fas fa-envelope"
            />
            <span
              class="admin-navbar__item-badge"
            >
              0
            </span>
          </span>
          <span
            class="admin-navbar__item-text"
          >
            Messages
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
